src/server/pkg/storage/fileset/merge.go

```diff
@@ -341,86 +317,66 @@ func newTagSetMergeReader(frs []*FileReader) *TagSetMergeReader {
 }
 
 // Iterate iterates over the tag merge readers in the merged tagset.
-func (tsmr *TagSetMergeReader) Iterate(f func(*TagMergeReader) error) error {
+func (tsmr *TagSetMergeReader) Iterate(cb func(*TagMergeReader) error) error {
 	return tsmr.pq.iterate(func(ss []stream, _ ...string) error {
-		// Convert generic stream to tag stream.
-		var tss []*tagStream
-		for _, s := range ss {
-			if _, ok := s.(*deleteTagStream); ok {
-				tss = nil
-				continue
+		return tsmr.mergeTag(ss, cb)
+	})
+}
+
+func (tsmr *TagSetMergeReader) mergeTag(ss []stream, cb func(*TagMergeReader) error, deleter ...func(string)) (retErr error) {
+	// Convert generic streams to tag streams.
+	var i int
+	var tss []*tagStream
+	for ; i < len(ss); i++ {
+		if dts, ok := ss[i].(*deleteTagStream); ok {
+			if len(deleter) > 0 {
+				deleter[0](dts.tag.Id)
```

Comment:
Same feedback as earlier. deleter should not be variadic. If it is nil, assign a function that always returned false.

Reply:
Same as above. It is an optional callback that performs a deletion, not a predicate.